src/test/java/com/puppycrawl/tools/checkstyle/filters/SuppressionXpathSingleFilterTest.java

```diff
@@ -0,0 +1,302 @@
+    }
+
+    @Test
+    public void testNonMatchingFileNameModuleIdAndCheck() throws Exception {
+        final String xpath = "NON_MATCHING_QUERY";
+        final SuppressionXpathSingleFilter filter =
+                createSuppressionXpathSingleFilter("InputSuppressionXpathSingleFilter",
+                        null, null, null, xpath);
+        final TreeWalkerAuditEvent ev = getEvent(3, 0,
+                TokenTypes.CLASS_DEF);
+        assertTrue("Event should be accepted", filter.accept(ev));
+    }
+
+    @Test
+    public void testNullModuleIdAndNonMatchingChecks() throws Exception {
+        final String xpath = "NON_MATCHING_QUERY";
+        final SuppressionXpathSingleFilter filter = createSuppressionXpathSingleFilter(
+                "InputSuppressionXpathSingleFilter", "NonMatchingRegexp", null, null, xpath);
+        final TreeWalkerAuditEvent ev = getEvent(3, 0,
+                TokenTypes.CLASS_DEF);
+        assertTrue("Event should be accepted", filter.accept(ev));
+    }
+
+    @Test
+    public void testDecideByMessage() throws Exception {
+        final LocalizedMessage message = new LocalizedMessage(0, 0,
+                TokenTypes.CLASS_DEF, "", "",
+                null, null, null, getClass(), "Test");
+        final TreeWalkerAuditEvent ev = new TreeWalkerAuditEvent(fileContents, file.getName(),
+                message, JavaParser.parseFile(file, JavaParser.Options.WITHOUT_COMMENTS));
+        final SuppressionXpathSingleFilter filter1 = createSuppressionXpathSingleFilter(
+                null, null, "Test", null, null);
+        final SuppressionXpathSingleFilter filter2 = createSuppressionXpathSingleFilter(
+                null, null, "Bad", null, null);
+        assertFalse("Message match", filter1.accept(ev));
+        assertTrue("Message not match", filter2.accept(ev));
+    }
+
+    @Test
+    public void testThrowException() {
+        final String xpath = "/CLASS_DEF[@text='InputSuppressionXpathSingleFilter']";
+        final SuppressionXpathSingleFilter filter =
+                createSuppressionXpathSingleFilter("InputSuppressionXpathSingleFilter",
+                        "Test", null, null, xpath);
+        final LocalizedMessage message =
+                new LocalizedMessage(3, 0, TokenTypes.CLASS_DEF, "",
+                        "", null, null, "id19",
+                        getClass(), null);
+        final TreeWalkerAuditEvent ev = new TreeWalkerAuditEvent(fileContents,
+                file.getName(), message, null);
+        try {
+            filter.accept(ev);
+            fail("Exception is expected");
+        }
+        catch (IllegalStateException ex) {
+            assertTrue("Exception message does not match expected one",
+                    ex.getMessage().contains("Cannot initialize context and evaluate query"));
+        }
+    }
+
+    private static SuppressionXpathSingleFilter createSuppressionXpathSingleFilter(
+            String... properties) {
```

Comment:
I have done this initially. However, CI told me that I got too much string parameters and should consider to create a container to store them.

Reply:
It is ok to suppress violations on test code.

Reply:
How to suppress this kind of violations on test code? Should I add something on it?

Reply:
@jack870131 Who was reporting violation? You need to know this in order to suppress because each tool has their own way to suppress.
Checkstyle uses suppression file. PMD has xpath in one of it's configuration files. TeamCity has `@noinspection` javadoc.

Reply:
> SuppressionXpathSingleFilterTest:281 Rule:UseObjectForClearerAPI Priority:3 Rather than using a lot of String arguments, consider using a container object for those values..


This is a PMD violation and must suppressed in https://github.com/checkstyle/checkstyle/blob/master/config/pmd-test.xml . Please put a reason above the suppression.

Reply:
Don't use `@SuppressWarnings("PMD")`. Suppression should be in file I mentioned.

Place something like:
````
    <rule ref="category/java/design.xml/UseObjectForClearerAPI" />
    <properties>
            <property name="violationSuppressXPath"
                      value="//ClassOrInterfaceDeclaration[@Image='SuppressionXpathSingleFilterTest']
                  //MethodDeclarator[@Image='createSuppressionXpathSingleFilter']"/>
        </properties>
    </rule>
````


Add your own comment about property specifying reason for suppression.